src/containers/Steps/StepThree/index.tsx

```diff
@@ -1,39 +1,104 @@
+import { useEffect, useState } from 'react';
+
 import StepWrapper from '../../../components/StepWrapper';
 
-import Copy from '../../../assets/Copy';
+import { useAppSelector } from '../../../hooks/useRedux';
 
-import { OptionType } from '../../../models';
 import copyText from '../../../utils/copyText';
+import generateQrCode from '../../../utils/generateQrCode';
+
+import { ITokenResponse, OptionType } from '../../../models';
+
+import Copy from '../../../assets/Copy';
 
 interface StepThreeProps {
   prevStep: () => void;
+  setStep: (_: number) => void;
   selectedNetwork: OptionType | null;
-  selectedToken: OptionType | null;
+  selectedToken: ITokenResponse | null;
+  setIsPayed: (_: boolean) => void;
 }
 
-const StepThree = ({ prevStep, selectedNetwork, selectedToken }: StepThreeProps) => {
+const StepThree = ({
+  prevStep,
+  selectedNetwork,
+  selectedToken,
+  setStep,
+  setIsPayed,
+}: StepThreeProps) => {
+  const [qrCode, setQrCode] = useState<string | null>(null);
+
+  const { data, loading } = useAppSelector((store) => store.data);
+
+  let amount = '0';
+  if (!loading && data) {
+    amount =
+      data.tokens.find((token) => token.token.symbol === selectedToken?.symbol)?.amount || '0';
+
+    if (data.status === 'completed') {
+      setIsPayed(true);
+      setStep(4);
+    }
+  }
+
+  const paymentAddress = import.meta.env.VITE_CONTRACT_ADDRESS;
+
+  const generateQr = async () => {
+    try {
+      const qr = await generateQrCode(paymentAddress);
+      setQrCode(qr);
+    } catch (error) {
+      console.error('Error generating QR code:', error);
+    }
+  };
```

Comment:
remove the function and put all the code in the useEffect directly as it only gets called once.